File: python/packages/kagent-adk/src/kagent/adk/_tool_pairing.py
"""Tool call/response pairing repair for the model request.

A tool call and its result are persisted as two separate session events. If a
turn ends between them (process restart, OOM, client disconnect, cancellation,
or a second message arriving while a slow tool is still running), the session
keeps a ``function_call`` with no matching ``function_response``.

History is replayed verbatim on every later turn, and providers that require
strict pairing reject the whole conversation. Anthropic returns
``tool_use ids were found without tool_result blocks immediately after`` and the
session stays unusable until it is deleted.

The repair runs against the model request, not the store: ADK builds the request
from deep copies of the session events (see ``flows/llm_flows/contents.py``), so
the recorded history stays intact for the UI and sessions already broken in the
field heal on their next turn without a migration.

Pairing is checked positionally, against the immediately following content,
because that is the invariant the provider enforces. A response that exists
somewhere else in the history does not satisfy it.

On a conversation whose calls are all answered this is a no-op, and any
conversation it does change is one the provider would have rejected outright.
"""

from __future__ import annotations

from google.adk.agents.callback_context import CallbackContext
from google.adk.models.llm_request import LlmRequest
from google.genai import types

# Stands in for a result that was never recorded. Deliberately states no cause:
# the call may have been interrupted, or may belong to a long-running tool that
# has not returned yet. Matches the placeholder the OpenAI and Go converters
# already use, so providers that repair on their own keep the same behaviour.
MISSING_TOOL_RESULT = "No response available for this function call."


def _call_ids(content: types.Content | None) -> set[str | None]:
    if content is None:
        return set()
    return {p.function_call.id for p in content.parts or [] if p.function_call}


def _response_ids(content: types.Content | None) -> set[str | None]:
    if content is None:
        return set()
    return {p.function_response.id for p in content.parts or [] if p.function_response}


def _drop_orphaned_responses(contents: list[types.Content]) -> list[types.Content]:
    """Remove function_response parts whose call is not in the preceding content."""
    kept: list[types.Content] = []
    for index, content in enumerate(contents):
        responses = _response_ids(content)
        if not responses:
            kept.append(content)
            continue

        answerable = _call_ids(contents[index - 1]) if index > 0 else set()
        parts = [p for p in content.parts or [] if not p.function_response or p.function_response.id in answerable]
        if not parts:
            continue
        content.parts = parts
        kept.append(content)
    return kept


def _synthesize_missing_responses(contents: list[types.Content]) -> list[types.Content]:
    """Give every function_call a function_response in the immediately following content."""
    repaired: list[types.Content] = []
    for index, content in enumerate(contents):
        repaired.append(content)

        calls = [p.function_call for p in content.parts or [] if p.function_call]
        if not calls:
            continue

        following = contents[index + 1] if index + 1 < len(contents) else None
        answered = _response_ids(following)
        missing = [call for call in calls if call.id not in answered]
        if not missing:
            continue

        parts = [
            types.Part(
                function_response=types.FunctionResponse(
                    id=call.id,
                    name=call.name,
                    response={"result": MISSING_TOOL_RESULT},
                )
            )
            for call in missing
        ]

        # Join an existing response turn so the results stay in one message;
        # otherwise the results need a turn of their own, before whatever
        # currently follows.
        if following is not None and answered:
            following.parts = list(following.parts or []) + parts
        else:
            repaired.append(types.Content(role="user", parts=parts))
    return repaired


def repair_tool_call_pairing_callback(
    callback_context: CallbackContext,
    llm_request: LlmRequest,
) -> None:
    """Before-model callback that pairs every tool call with a tool result.

    Drops a result whose call is gone, then supplies a placeholder result for a
    call that has none, so the request satisfies the strict call/result pairing
    that Anthropic (and Bedrock) require.
    """
    if not llm_request.contents:
        return None
    contents = _drop_orphaned_responses(list(llm_request.contents))
    llm_request.contents = _synthesize_missing_responses(contents)
    return None

File: go/adk/pkg/agent/tool_pairing.go
package agent

import (
	"github.com/go-logr/logr"
	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/agent/llmagent"
	adkmodel "google.golang.org/adk/v2/model"
	"google.golang.org/adk/v2/session"
	"google.golang.org/genai"
)

// missingToolResult stands in for a result that was never recorded. It states
// no cause: the call may have been interrupted, or the process may have died
// between the two events. It matches the placeholder the Anthropic and OpenAI
// converters already use, so a request that reaches one of those unchanged
// behaves exactly as before.
const missingToolResult = "No response available for this function call."

// pendingToolResult stands in for a call ADK is deliberately holding open: a
// long-running tool, a human approval, or ask_user. No function response exists
// for these until the answer arrives, so the call is pending rather than lost.
// The distinction matters: told a tool "returned nothing", a model reissues it
// or proceeds; told the call is still awaiting a response, it can wait.
const pendingToolResult = "This call is awaiting a response and has not completed yet."

// MakeToolPairingCallback repairs tool call/response pairing in the model
// request.
//
// A tool call and its result are persisted as two separate session events. If a
// turn ends between them (process restart, OOM, client disconnect,
// cancellation), the session keeps a function call with no matching response.
// History is replayed on every later turn, and providers that require strict
// pairing reject the whole conversation, leaving the session unusable until it
// is deleted.
//
// The repair runs against the request rather than the store, so recorded
// history stays intact and sessions already broken heal on their next turn. It
// also sits above the session service, so it applies whichever store backs the
// session.
//
// ADK hands each request a deep copy of the content, so assigning Parts here
// cannot reach persisted history.
//
// Pairing is checked positionally, against the immediately following content,
// because that is the invariant the provider enforces.
//
// A conversation whose calls are all answered is left untouched, and any
// conversation this does change is one the provider would have rejected.
func MakeToolPairingCallback(log logr.Logger) llmagent.BeforeModelCallback {
	return func(ctx agent.Context, req *adkmodel.LLMRequest) (*adkmodel.LLMResponse, error) {
		if len(req.Contents) == 0 {
			return nil, nil
		}
		var source sessionSource
		if ctx != nil {
			source = ctx
		}
		repaired, synthesized := synthesizeMissingResponses(req.Contents, pendingCallIDs(source))
		if synthesized > 0 {
			// Logged because a repaired request means a turn ended between a
			// tool call and its result somewhere upstream.
			log.Info("Paired unanswered tool calls before the model request", "count", synthesized)
		}
		req.Contents = repaired
		return nil, nil
	}
}

// sessionSource is the part of agent.Context this file needs.
type sessionSource interface {
	Session() session.Session
}

// pendingCallIDs returns the ids of calls ADK is holding open for a
// long-running tool or an approval. They are read from the session events
// because LongRunningToolIDs lives on the event and does not survive the
// conversion to contents.
func pendingCallIDs(source sessionSource) map[string]bool {
	pending := map[string]bool{}
	if source == nil {
		return pending
	}
	current := source.Session()
	if current == nil {
		return pending
	}
	events := current.Events()
	if events == nil {
		return pending
	}
	for event := range events.All() {
		if event == nil {
			continue
		}
		for _, id := range event.LongRunningToolIDs {
			pending[id] = true
		}
	}
	return pending
}

// synthesizeMissingResponses gives every function call a function response in
// the immediately following content, and reports how many it had to supply.
func synthesizeMissingResponses(contents []*genai.Content, pending map[string]bool) ([]*genai.Content, int) {
	synthesized := 0
	repaired := make([]*genai.Content, 0, len(contents))
	for index, content := range contents {
		if content == nil {
			continue
		}
		repaired = append(repaired, content)

		calls := functionCalls(content)
		if len(calls) == 0 {
			continue
		}

		var following *genai.Content
		if index+1 < len(contents) {
			following = contents[index+1]
		}
		answered := responseIDs(following)
		missing := unanswered(calls, answered)
		if len(missing) == 0 {
			continue
		}

		synthesized += len(missing)
		parts := make([]*genai.Part, 0, len(missing))
		for _, call := range missing {
			parts = append(parts, &genai.Part{
				FunctionResponse: &genai.FunctionResponse{
					ID:       call.ID,
					Name:     call.Name,
					Response: map[string]any{"result": placeholderFor(call, pending)},
				},
			})
		}

		// Join an existing response turn so the results stay in one message;
		// otherwise the results need a turn of their own, before whatever
		// currently follows.
		if following != nil && len(answered) > 0 {
			following.Parts = append(following.Parts, parts...)
			continue
		}
		repaired = append(repaired, &genai.Content{Role: genai.RoleUser, Parts: parts})
	}
	return repaired, synthesized
}

func placeholderFor(call *genai.FunctionCall, pending map[string]bool) string {
	if call.ID != "" && pending[call.ID] {
		return pendingToolResult
	}
	return missingToolResult
}

// unanswered returns the calls with no response in answered. Ids are consumed
// one at a time rather than matched through a set, so a turn carrying several
// calls with no id (Gemini omits them, and ADK strips its own "adk-" ids before
// the request is built) does not have one response silently answer all of them.
func unanswered(calls []*genai.FunctionCall, answered []string) []*genai.FunctionCall {
	remaining := make([]string, len(answered))
	copy(remaining, answered)

	var missing []*genai.FunctionCall
	for _, call := range calls {
		matched := false
		for i, id := range remaining {
			if id == call.ID {
				remaining = append(remaining[:i], remaining[i+1:]...)
				matched = true
				break
			}
		}
		if !matched {
			missing = append(missing, call)
		}
	}
	return missing
}

func functionCalls(content *genai.Content) []*genai.FunctionCall {
	if content == nil {
		return nil
	}
	var calls []*genai.FunctionCall
	for _, part := range content.Parts {
		if part != nil && part.FunctionCall != nil {
			calls = append(calls, part.FunctionCall)
		}
	}
	return calls
}

func responseIDs(content *genai.Content) []string {
	if content == nil {
		return nil
	}
	var ids []string
	for _, part := range content.Parts {
		if part != nil && part.FunctionResponse != nil {
			ids = append(ids, part.FunctionResponse.ID)
		}
	}
	return ids
}
